File: uutuuskirjat/qa_verify.js

const https = require('https');

// Utils
function fetchJson(url) {
    return new Promise((resolve, reject) => {
        https.get(url, (res) => {
            let body = '';
            res.on('data', chunk => body += chunk);
            res.on('end', () => {
                try {
                    resolve(JSON.parse(body));
                } catch (e) { resolve({}); }
            });
        }).on('error', reject);
    });
}

const CLEAN_AUTHOR = (author) => {
    if (author && author.includes(',')) {
        const parts = author.split(',');
        if (parts.length === 2) {
            return `${parts[1].trim()} ${parts[0].trim()}`;
        }
    }
    return author;
};

async function testBook(title, rawAuthor) {
    console.log(`\n--- TESTING: ${title} ---`);
    console.log(`Raw Author: ${rawAuthor}`);
    const clean = CLEAN_AUTHOR(rawAuthor);
    console.log(`Clean Author: ${clean}`);

    // 1. Google Cover Search (Strategy B logic)
    const q = `intitle:${encodeURIComponent(title)}+inauthor:${encodeURIComponent(clean)}`;
    const url = `https://www.googleapis.com/books/v1/volumes?q=${q}&maxResults=1`;

    console.log(`Fetching Google Cover (Strategy B)...`);
    const data = await fetchJson(url);

    if (data.items && data.items.length > 0) {
        const info = data.items[0].volumeInfo;
        console.log(`[SUCCESS] Found Google Book!`);
        console.log(`  Title: ${info.title}`);
        console.log(`  Authors: ${info.authors ? info.authors.join(', ') : '?'}`);
        console.log(`  ImageLinks: ${info.imageLinks ? 'YES' : 'NO'}`);
        console.log(`  ISBN: ${info.industryIdentifiers ? JSON.stringify(info.industryIdentifiers) : 'None'}`);
    } else {
        console.log(`[FAILURE] No results found from Google with clean author.`);

        // Try loose
        console.log(`Retrying Loose Search...`);
        const qLoose = `${encodeURIComponent(title)}+${encodeURIComponent(clean)}`;
        const urlLoose = `https://www.googleapis.com/books/v1/volumes?q=${qLoose}&maxResults=1`;
        const dataLoose = await fetchJson(urlLoose);
        if (dataLoose.items && dataLoose.items.length > 0) {
            console.log(`[SUCCESS] Found via Loose Search!`);
            console.log(`  Title: ${dataLoose.items[0].volumeInfo.title}`);
        } else {
            console.log(`[FAILURE] Loose search also failed.`);
        }
    }
}

async function run() {
    await testBook("Talonvahti", "Sager, Riley");
    await testBook("Sokeiden valtakunta", "Penny, Louise");
    await testBook("Mestaritontun seikkailut", "Kunnas, Mauri"); // Control
}

run();

File: kirjavinkit/qa_verify.py
import urllib.request
import urllib.parse
import json

def fetch_json(url):
    try:
        with urllib.request.urlopen(url) as response:
            return json.loads(response.read().decode())
    except Exception as e:
        print(f"Error fetching {url}: {e}")
        return {}

def clean_author_name(author):
    if author and ',' in author:
        parts = author.split(',')
        if len(parts) == 2:
            return f"{parts[1].strip()} {parts[0].strip()}"
    return author

def test_book(title, raw_author):
    print(f"\n--- TESTING: {title} ---")
    print(f"Raw Author: {raw_author}")
    clean = clean_author_name(raw_author)
    print(f"Clean Author: {clean}")

    # 1. Google Cover Search (Strategy B logic)
    q = f"intitle:{urllib.parse.quote(title)}+inauthor:{urllib.parse.quote(clean)}"
    url = f"https://www.googleapis.com/books/v1/volumes?q={q}&maxResults=1"
    
    print("Fetching Google Cover (Strategy B)...")
    data = fetch_json(url)
    
    if data.get('items'):
        info = data['items'][0].get('volumeInfo', {})
        print("[SUCCESS] Found Google Book!")
        print(f"  Title: {info.get('title')}")
        print(f"  Authors: {', '.join(info.get('authors', []))}")
        print(f"  ImageLinks: {'YES' if info.get('imageLinks') else 'NO'}")
        print(f"  ISBN: {json.dumps(info.get('industryIdentifiers', []))}")
    else:
        print("[FAILURE] No results found from Google with clean author.")
        
        # Try loose
        print("Retrying Loose Search...")
        q_loose = f"{urllib.parse.quote(title)}+{urllib.parse.quote(clean)}"
        url_loose = f"https://www.googleapis.com/books/v1/volumes?q={q_loose}&maxResults=1"
        data_loose = fetch_json(url_loose)
        if data_loose.get('items'):
             print("[SUCCESS] Found via Loose Search!")
             print(f"  Title: {data_loose['items'][0]['volumeInfo'].get('title')}")
        else:
             print("[FAILURE] Loose search also failed.")

def run():
    test_book("Talonvahti", "Sager, Riley")
    test_book("Sokeiden valtakunta", "Penny, Louise")
    test_book("Mestaritontun seikkailut", "Kunnas, Mauri")

if __name__ == "__main__":
    run()
